### ISSUE TITLE
Unable to subscribe to Geyser events in Spigot plugin

### ISSUE THREAD

**Issue by ImDaBigBoss** [GeyserMC/Geyser#3694]
### Describe the bug

When I subscribe to a Geyser event using the `@Subscribe` annotation, events aren't actually sent.

### To Reproduce

1. Create a class in your favorite Spigot/Paper plugin and implement `EventRegistrar`.
2. Call `GeyserApi.api().eventBus().register(this, this);` in the constructor to register this new class as an event listener.
3. Add a few events:
```java
@Subscribe
public void onGeyserPostInitializeEvent(GeyserPostInitializeEvent event) {
    yourPluginInstance.getLogger().info("Oh hey, Geyser started up");
}
```
4. In the `onLoad()` function of your favorite plugin's main class, create an instance of your listener. I think it should be possible in `onEnable()` now, but I've got it there just to be safe.
5. Set Geyser as a dependency of your favorite spigot plugin inside `plugin.yml` like so: `depend: ["Geyser-Spigot"]`.

### Expected behaviour

I should see "Oh hey, Geyser started up" in the console, but I do not because the event is never fired.

### Server Version and Plugins

Paper version git-Paper-516 (MC: 1.19.4) (Implementing API version 1.19.4-R0.1-SNAPSHOT) (Git: f7717c3)

### Geyser Version

version 2.1.0-SNAPSHOT (git-master-0f99abc)

### Additional Context

I was however able to fix the problem, by calling `GeyserApi.api().eventBus().subscribe(this, WhateverTheEvent.class, this::myFunction);` after the event listener was registered. That isn't required in extensions, and shouldn't really be in other plugins either since the subscribe annotation should do it on its own.

**Comment by Complex1t** [comment]
Hi, are you able to describe more about how you were able to fix this problem?

I have followed your guidelines, and I have this class. It is based on the [TemplateGeyserExtension](https://github.com/GeyserCustomModelData/TemplateGeyserExtension) project, but I have adapted your steps as the `@Subscribe` annotation doesn't show as valid.
[https://pastebin.com/vffNTSne](https://pastebin.com/vffNTSne)

However, I'm getting a compiler error saying `Inferred type 'T' for type parameter 'T' is not within its bound; should implement 'org.geysermc.event.Event'`.

**Comment by Redned235** [comment]
`EventRegistrar` is meant to be implemented in the main class of whatever is using the event system (for plugins, this should be the plugin main class). As we cannot guarantee a `Plugin` instance when it comes to the extensions system, the API requires you implement `EventRegistrar` in your plugin's main class and you plug that in when calling the EventBus register method.

**Comment by Complex1t** [comment]
Thank you for the clarifications. I did not realize extensions are a whole separate system from plugins.

**Comment by onebeastchris** [comment]
Can confirm this also to be the case with Velocity plugins.
I'm wondering if the annotation processor is just missing? Even then, no idea why it would work on bungee or fabric, but not paper/velocity.

**Comment by Konicai** [comment]
Unable to reproduce with: https://github.com/kejonaMC/CrossplatForms/compare/main...api-test
https://api.mclo.gs/1/raw/auie5iq

The 
```[19:47:44 INFO]: [Geyser-Spigot] [STDOUT] a
[19:47:44 INFO]: [Geyser-Spigot] [STDOUT] b
[19:47:44 INFO]: [Geyser-Spigot] [STDOUT] c
```
is where the subscriber was successfully registered. And you can see the event handler do print a message at the end of the log

**Comment by ofunny** [comment]
I can confirm that bug on Velocity 
- Velocity 3.2.0-SNAPSHOT (git-19abb909-b265)
- Geyser version 2.2.0-SNAPSHOT (git-master-7a28ef5)

Simple test code:

```
(...)
        dependencies = {
            @Dependency(id = "geyser", optional = true)
        }
)
public class OFunnyWorld implements EventRegistrar {

    private final ProxyServer proxyServer;
    private final Logger logger;
    private final Path dataDirectory;

    @Inject
    public OFunnyWorld(ProxyServer proxyServer, Logger logger, @DataDirectory Path dataDirectory) {
        this.proxyServer = proxyServer;
        this.logger = logger;
        this.dataDirectory = dataDirectory;
    }// end OFunnyWorld

    @Subscribe
    public void onProxyInitialization(ProxyInitializeEvent event) {
        
        System.out.println("################# Starting to check for Geyser");
        // Testing Geyser login events to get a more accurate one on the paper servers
        if(proxyServer.getPluginManager().isLoaded("geyser")) {
            GeyserApi geyserAPI = GeyserApi.api();
            if(geyserAPI != null) {
                System.out.println("################# Registering event.");
                geyserAPI.eventBus().register(this, new GeyserEventListener());
            }// end if GeyserApi
        }// end if Geyser loaded

    }// end onProxyInitialization

}// end class OFunnyWorld

```
and 

```
public class GeyserEventListener {

    public GeyserEventListener() {
        System.out.println("################# Listener class created!");
    }

    @Subscribe(postOrder = PostOrder.NORMAL)
    public void onGeyserPostInitializeEvent(GeyserPostInitializeEvent event) {
        System.out.println("################# Geyser started!");
    }

    @Subscribe(postOrder = PostOrder.NORMAL)
    public void onGeyserSessionJoinEvent(SessionJoinEvent event) {
        System.out.println("################# Joined the session: "+event.connection().bedrockUsername());
    }

}// end GeyserEventListener

```
are not firing, will work around this by falling back to the consumer subscription approach for now.

**Comment by Konicai** [comment]
share a dump as well please

**Comment by ofunny** [comment]
> share a dump as well please
Certainly: https://dump.geysermc.org/M1Deiu3l6LrMr6DAJYbfACGvmXYItzne

**Comment by ofunny** [comment]
If anyone stumbles across this and needs an interim solution close to registering an EventListener class, you may use the following example until this bug might get fixed.

```
(your imports and your plugin information goes here ...)
@Dependency(id = "geyser", optional = true)
        }
)
public class OFunnyWorld implements EventRegistrar {

    private final ProxyServer proxyServer;
    private final Logger logger;
    private final Path dataDirectory;

    @Inject
    public OFunnyWorld(ProxyServer proxyServer, Logger logger, @DataDirectory Path dataDirectory) {
        this.proxyServer = proxyServer;
        this.logger = logger;
        this.dataDirectory = dataDirectory;

    }// end OFunnyWorld

    @Subscribe
    public void onProxyInitialization(ProxyInitializeEvent event) {
        
        if(proxyServer.getPluginManager().isLoaded("geyser")) new GeyserEventListener(this);

    }// end onProxyInitialization

}// end class OFunnyWorld

```

and 

```
(your imports here … )

public class GeyserEventListener {

    public GeyserEventListener(OFunnyWorld oFunnyWorld) {

        GeyserApi geyserAPI = GeyserApi.api();
        if(geyserAPI == null) return;

        // Register events here instead of @Subscribe
        geyserAPI.eventBus().subscribe(oFunnyWorld, GeyserPostInitializeEvent.class, this::onGeyserPostInitializeEvent);
        geyserAPI.eventBus().subscribe(oFunnyWorld, SessionJoinEvent.class, this::onGeyserSessionJoinEvent);

    }// end GeyserEventListener

    //@Subscribe(postOrder = PostOrder.NORMAL)
    public void onGeyserPostInitializeEvent(GeyserPostInitializeEvent event) {
        System.out.println("################# Geyser started!");
    }

    //@Subscribe(postOrder = PostOrder.NORMAL)
    public void onGeyserSessionJoinEvent(SessionJoinEvent event) {
        System.out.println("################# Joined the session: "+event.connection().bedrockUsername());
    }

}// end GeyserEventListener

```

**Comment by TBYT** [comment]
This issue is caused by something in the floodgate spigot jar. If you remove the jar entirely then it works, as you can test with the GeyserPostInitializeEvent. oFunny's solution of manually subscribing also circumvents the issue instead of having to delete the floodgate jar.

**Comment by ImDaBigBoss** [comment]
I'm not sure I see why Floodgate would do this, my specific server setup doesn't utilize Floodgate at all.

**Comment by onebeastchris** [comment]
After some debugging, i've found out that this is caused by our Events lib that doesn't check the correct class.
https://github.com/onebeastchris/Events/blob/master/src/main/java/org/geysermc/event/bus/impl/BaseBusImpl.java#L119 is supposed to return the current class (the one that you are supposed to register with , but it instead returns `java.lang.Class` in most cases. It's supposed to then iterate through the super classes of said class to eventually find your class, but this doesn't always occur as [this](https://github.com/onebeastchris/Events/blob/master/src/main/java/org/geysermc/event/bus/impl/BaseBusImpl.java#L150) instead returned object.class, which broke the loop (on purpose).

In short; this is an event lib issue that should hopefully be resolved quicker with the actual cause found; for now, subscribing to events manually works.

**Comment by ImDaBigBoss** [comment]
I'm pretty sure this is no longer the case. I can't think of any reason why this would stop happening, but since I don't face the issue anymore, I can't justify this issue.
